nexus-writer/src/main.rs

```diff
@@ -176,86 +155,172 @@ async fn main() -> Result<()> {
     }
 }
 
-fn process_digitizer_event_list_message(nexus: &mut Nexus, payload: &[u8]) {
-    match root_as_digitizer_event_list_message(payload) {
-        Ok(data) => match GenericEventMessage::from_digitizer_event_list_message(data) {
-            Ok(event_data) => match nexus.process_message(&event_data) {
-                Ok(run) => {
-                    if let Some(run) = run {
-                        let cur_span = tracing::Span::current();
-                        run.span().get().unwrap().in_scope(|| {
-                            let span = trace_span!("Digitiser Events List");
-                            span.follows_from(cur_span);
-                        });
+// Handles Run Span for
+macro_rules! link_current_span_to_run_span {
+    ($run:ident, $span_name:literal) => {
+        let cur_span = tracing::Span::current();
+        match $run.span().get() {
+            Ok(run_span) => run_span.in_scope(|| {
+                trace_span!($span_name).follows_from(cur_span);
+            }),
+            Err(e) => debug!("No run found. Error: {e}"),
+        }
+    };
+}
+
+impl NexusEngine {
```

Comment:
Why is this `impl` here and not in `engine.rs`?

Reply:
This impl extends the impl block in engine.rs, as it creates the spans which collects different FrameEventLists into a single trace, and it handles the context propagation using functions from OtelTracer. I've tried to keep references to OtelTracer exclusively in the main.rs file wherever possible.

Reply:
> I've tried to keep references to OtelTracer exclusively in the main.rs file wherever possible.

I'm not sure this really does much for readability over having `impl` blocks near the `struct` they refer to.

Moving it would also keep the size of `main.rs` down, right now it is starting to get large.

Reply:
The engine.rs file is even larger. I think it makes most sense to convert these from methods to functions and leave them in main.rs. We can probably compact it at a later date using macros.